nexus/types/src/deployment.rs

```diff
@@ -271,31 +300,231 @@ impl<'a> fmt::Display for BlueprintDisplay<'a> {
         writeln!(f, "internal DNS version: {}", b.internal_dns_version)?;
         writeln!(f, "comment: {}", b.comment)?;
         writeln!(f, "zones:\n")?;
-        for (sled_id, sled_zones) in &b.omicron_zones {
+
+        for (sled_id, sled_zones) in &b.blueprint_zones {
             writeln!(
                 f,
                 "  sled {}: Omicron zones at generation {}",
                 sled_id, sled_zones.generation
             )?;
             for z in &sled_zones.zones {
-                writeln!(
-                    f,
-                    "    {} {} {}",
-                    z.id,
-                    if b.zones_in_service.contains(&z.id) {
-                        "in service    "
-                    } else {
-                        "not in service"
-                    },
-                    z.zone_type.label(),
-                )?;
+                writeln!(f, "    {}", z.display())?;
             }
         }
 
         Ok(())
     }
 }
 
+/// Information about an Omicron zone as recorded in a blueprint.
+///
+/// Currently, this is similar to [`OmicronZonesConfig`], but also contains a
+/// per-zone [`BlueprintZoneDisposition`].
+///
+/// Part of [`Blueprint`].
+#[derive(Debug, Clone, Eq, PartialEq, JsonSchema, Deserialize, Serialize)]
+pub struct BlueprintZonesConfig {
+    /// Generation number of this configuration.
+    ///
+    /// This generation number is owned by the control plane. See
+    /// [`OmicronZonesConfig::generation`] for more details.
+    pub generation: Generation,
+
+    /// The list of running zones.
+    pub zones: Vec<BlueprintZoneConfig>,
+}
+
+impl BlueprintZonesConfig {
+    /// Constructs a new [`BlueprintZonesConfig`] from a collection's zones.
+    ///
+    /// For the initial blueprint, all zones within a collection are assumed to
+    /// be in-service.
+    pub fn initial_from_collection(collection: &OmicronZonesConfig) -> Self {
+        let zones = collection
+            .zones
+            .iter()
+            .map(|z| BlueprintZoneConfig {
+                config: z.clone(),
+                disposition: BlueprintZoneDisposition::InService,
+            })
+            .collect();
+
+        let mut ret = Self {
+            // An initial `BlueprintZonesConfig` reuses the generation from
+            // `OmicronZonesConfig`.
+            generation: collection.generation,
+            zones,
+        };
+        ret.sort();
+
+        ret
+    }
+
+    /// Sorts the list of zones stored in this configuration.
+    ///
+    /// This is not strictly necessary. But for testing, it's helpful for
+    /// things to be in sorted order.
```

Comment:
Ah, I see the comment moved to here. I think it may make more sense at the call site? It's possible (although unlikely and I hope we don't) we could introduce some _necessary_ `.sort()` call at some point in the future.

Reply:
Left the comment in for now -- hopefully anyone adding the future necessary `sort()` will see the comment.

In #5270 I'm actually going to change the sort order to (zone kind, UUID), so all zones with the same kind are grouped together in output. That sort order may not make sense for non-test use cases.